<font size = "5"> **Chapter 2: [Diffraction](CH2_00-Diffraction.ipynb)** </font>


<hr style="height:1px;border-top:4px solid #FF8200" />



# The Electron


[Download](https://raw.githubusercontent.com/gduscher/MSE672-Introduction-to-TEM//main/Diffraction/CH2_01-Electron.ipynb)

[![OpenInColab](https://colab.research.google.com/assets/colab-badge.svg)](
    https://colab.research.google.com/github/gduscher/MSE672-Introduction-to-TEM/blob/main//Diffraction/CH2_01-Electron.ipynb)

part of

<font size = "5"> **[MSE672:  Introduction to Transmission Electron Microscopy](../_MSE672_Intro_TEM.ipynb)**</font>

**Spring 2026**<br>
by Gerd Duscher

Microscopy Facilities<br>
Institute of Advanced Materials & Manufacturing<br>
Materials Science & Engineering<br>
The University of Tennessee, Knoxville


Background and methods to analysis and quantification of data acquired with transmission electron microscopes.



First we load the code to make figures from pyTEMlib
## Import packages for figures and
### Check Installed Packages

In [1]:
import sys
import importlib.metadata
def test_package(package_name):
    """Test if package exists and returns version or -1"""
    try:
        version = importlib.metadata.version(package_name)
    except importlib.metadata.PackageNotFoundError:
        version = '-1'
    return version

if test_package('pyTEMlib') < '0.2026.1.0':
    print('installing pyTEMlib')
    !{sys.executable} -m pip install  --upgrade pyTEMlib -q

print('done')

done


Please restart the session (in Runtime menu) in Google Colab.

### Load the plotting and figure packages

In [1]:
%matplotlib widget
import matplotlib.pylab as plt
import numpy as np
import sys
if 'google.colab' in sys.modules:
    from google.colab import output
    output.enable_custom_widget_manager()

import scipy  # we will use the constants part only
import pyTEMlib

## Interaction of Common Particles with Matter

We generally use electron, photons, and neutrons for diffraction/scattering
experiments.

These particles interact with differently with matter:

 <table style="width:80%">

  <tr>
    <td>X-rays</td>
    <td>$\leftrightarrow$</td>
    <td>electron density</td>
  </tr>
  <tr>
    <td>neutrons</td>
    <td>$\leftrightarrow$</td>
    <td>mass of nucleus</td>
  </tr>
  <tr>
    <td>neutrons</td>
    <td>$\leftrightarrow$</td>
    <td>magnetic moment</td>
  </tr>
  <tr>
    <td>electrons</td>
    <td>$\leftrightarrow$</td>
    <td>screened charge of nucleus</td>
  </tr>

</table>

We will deal with the nature of electrons more closely in the following

## Non-relativistic de Broglie Wavelength


The electron is a elementary particle with spin $\frac{1}{2}$ (lepton).


**Non--relativistic De Broglie wavelength** of electron:

$\lambda = \frac{h}{p} = \frac{h}{\sqrt{2m_0E_{kin}}} \approx \frac{1.22}{\sqrt{E_{kin}}}$


E is the kinetic energy of the electron: $E_{kin} = eU $ [eV].

The wave length in a TEM is usually
a couple of picometers . This is a
factor of 100 smaller than your
XRD-source.

Obvioulsy, we are in the wave picture right now.

In [2]:
# --------- input  ---------------------------
acceleration_voltage_V = U = 100.0 * 1000.0 #V
# --------------------------------------------
## energy
E_kin = acceleration_voltage_V * scipy.constants.elementary_charge
h = scipy.constants.Planck
m0 = scipy.constants.electron_mass
c = scipy.constants.speed_of_light

wave_length_m = h/np.sqrt(2*m0*E_kin) # non-relativistic wavelength in m

# # please note that we will keep all length units in Angstrom if possible.
# # otherwise we use only SI units!!!
wave_length_A = wave_length_m *1e10 # now in Angstrom

print(f'Classic wave length is {wave_length_A*100.:.2f} pm for acceleration voltage {acceleration_voltage_V/1000.:.1f} kV')
# Notice that we change units in the output to make them most readable.

print(f' which is a velocity of {np.sqrt(2/m0*E_kin):.2f} m/s or {np.sqrt(2/m0*E_kin)/c*100:.2f}% of the speed of light')

Classic wave length is 3.88 pm for acceleration voltage 100.0 kV
 which is a velocity of 187553726.08 m/s or 62.56% of the speed of light


## Relativistic Correction
In the table below we see that the speeds of the electron is rather close to the speed of light $c$

The formula for relativistic corrected wavelength is:
$\lambda = \frac{h}{\sqrt{2m_e E_{kin} *(1+\frac{E_{kin}}{2 m_e c^2})}}$

**Please note:** All units are internally in SI units: kg, s, V, J, except the length wihich is in nm!

We multiply with the appropriate factors for the output

In [3]:
# ---------Input: Acceleration Voltage --
E0 = acceleration_voltage_V = 100.0 *1000.0 # V
# ---------------------------------------
E_kin = acceleration_voltage_V * scipy.constants.elementary_charge
h = scipy.constants.Planck
m0 = scipy.constants.electron_mass
c = scipy.constants.speed_of_light

#relativisitic wavelength
wave_length = h/np.sqrt(2* m0 * E_kin * (1 + E_kin / (2 * m0 * c**2))) #in m

print(f'The relativistically corrected wave length is {wave_length*1e12:.2f} pm for acceleration voltage {acceleration_voltage_V/1000:.1f} kV')

The relativistically corrected wave length is 3.70 pm for acceleration voltage 100.0 kV


100kV : $\lambda$ = 4 pm $<$ than diameter of an atom

The relativistic parameters are:


|E (keV)|$\lambda$ (pm) | M/m$_0$ | v/c|
--------|---------------|---------|----|
|10 |  12.2 | 1.0796 | 0.1950 |
|30 | 6.98 | 1.129 | 0.3284 |
|100 | 3.70 | 1.1957 | 0.5482|
|200 | 2.51 | 1.3914 |  0.6953|
|400 | 1.64 | 1.7828 | 0.8275 |
|1000 | 0.87 | 2.9569 | 0.9411|

The same functionality (and code) is used in the kinematic_scattering library and we can test the values of above table.

Please change the acceleration voltage (**acceleration_voltage_V**) above.



### Relativistic velocity

$$\frac{v^2}{c^2} = \frac{E_{kin}(E_{kin}+2m_e c^2)}{(E_{kin}+m_e c^2)^2}$$

In [4]:
v = np.sqrt(E_kin * (E_kin + 2 * m0 * c**2)/(E_kin + m0 * c**2) **2)*c

print(f'For an acceleration voltage of {acceleration_voltage_V/1000:.0f}keV: ')
print(f'The classic velocity of the electron  is {np.sqrt(2/m0 * E_kin):.2f} m/s or {np.sqrt(2 / m0 * E_kin)/c * 100:.2f}% of the speed of light')
print(f'The relativistic velocity of the electron  is {v:.2f} m/s or {v/c*100:.2f}% of the speed of light')

For an acceleration voltage of 100keV: 
The classic velocity of the electron  is 187553726.08 m/s or 62.56% of the speed of light
The relativistic velocity of the electron  is 164352479.65 m/s or 54.82% of the speed of light


## That means that the resolution is not limited by the wavelength!
The formula for relativistic corrected wavelength is:

$\lambda = \frac{h}{\sqrt{2m_e E_{kin} *(1+\frac{E_{kin}}{2 m_e c^2})}}$

In [9]:
# ----- Input -----
acceleration_voltage= 30 * 1000.0
# -------------------

wave_length = pyTEMlib.diffraction_tools.get_wavelength(acceleration_voltage, unit='A')
print(f'The relativistically corrected wave length is {wave_length*1e2:.2f} pm for acceleration voltage {acceleration_voltage/1000:.1f} kV')

# Wavelength in Angstrom
def get_wavelength(acceleration_voltage):
    """
    Calculates the relativistic corrected de Broglie wave length of an electron

    Input:
    ------
        acceleration voltage in volt
    Output:
    -------
        wave length in Angstrom
    """
    E = acceleration_voltage * scipy.constants.elementary_charge
    h = scipy.constants.Planck
    m0 = scipy.constants.electron_mass
    c = scipy.constants.speed_of_light

    wavelength = h / np.sqrt(2 * m0 * E * (1 + (E / (2 * m0 * c ** 2))))
    return wavelength * 10**10


The relativistically corrected wave length is 1.97 pm for acceleration voltage 300.0 kV


In [10]:
help(pyTEMlib.diffraction_tools.get_wavelength)

Help on function get_wavelength in module pyTEMlib.utilities:

get_wavelength(acceleration_voltage: float, unit: str = 'm') -> float
    Calculates the relativistic corrected de Broglie wavelength of an electron in meter
    and converts it to the desired unit (m, mm, μm, nm, A, Å, pm)

    Parameter:
    ---------
    acceleration_voltage: float
        acceleration voltage in volt
    unit: str
        unit of the wavelength, default is meter ('m')
    Returns:
    -------
    wavelength: float
        wave length in units given, default meter



In [11]:
help(pyTEMlib.diffraction_tools)

Help on package pyTEMlib.diffraction_tools in pyTEMlib:

NAME
    pyTEMlib.diffraction_tools

DESCRIPTION
    diffraction tools subpackage
    A collection of tools to analyze diffraction data
    part of pyTEMlib
    author: Gerd Duscher, UTK

PACKAGE CONTENTS
    basic
    diffraction_plot
    dynamic
    kinematic

FUNCTIONS
    calculate_holz(dif)
        Calculate HOLZ lines (of allowed reflections)

    calculate_kikuchi(atoms, tags, verbose)
        Calculate Kikuchi lines (of allowed reflections)

    calculate_laue_zones(atoms, tags, verbose)
        Calculate Laue Zones (of allowed reflections)

        ###########################
        Below is the expression given in most books.
        However, that would only work for orthogonal crystal systems
        Laue_Zone = abs(np.dot(hkl_allowed,tags['zone_hkl']))
        works only for orthogonal systems

        Below expression works for all crystal systems
        Remember we have already tilted, and so the dot product is tr

## Particle Flux and Current

It is important todetermine the order of magitude of how many electrons are hitting the sample.

The electron sources deliver in the order of $\mu$A current, but most of these electrons are not used.

In a modern electron microscope, we talk about a range of 1pA to 1nA in the electron beam.

We start with the defition of an Ampere:
$$A = \frac{C}{s}$$

That definition is enough to calculate the number ofelectron per time unit (flux).

In [12]:
print(f" elementary charge: {scipy.constants.physical_constants['elementary charge'][0]:.5g} {scipy.constants.physical_constants['elementary charge'][1]}")
print(f'\n  1pA is {1e-12/scipy.constants.elementary_charge:.3} electrons/s or {1e-12/scipy.constants.elementary_charge/1000:.0f} electrons/ms')
print(f' 10pA is {10e-12/scipy.constants.elementary_charge:.3} electrons/s')
print(f'100pA is {100e-12/scipy.constants.elementary_charge*1 :.3} electrons/s')

print(f'\n at 10pA an electron will hit the sample every {scipy.constants.elementary_charge/10e-12 * 1e9:.2f} ns ')

 elementary charge: 1.6022e-19 C

  1pA is 6.24e+06 electrons/s or 6242 electrons/ms
 10pA is 6.24e+07 electrons/s
100pA is 6.24e+08 electrons/s

 at 10pA an electron will hit the sample every 16.02 ns 


We see that we have much lower fluence in the TEM than in a laser (how could they do femtosecond pulses otherwise).


### Question
How long does one have to integrate a detector to register 1 electron with an electron beam of 1.6pA?

## Navigation
- <font size = "3">  **Back Chapter 1: [Introduction](../Introduction/CH1_00-Introduction.ipynb)** </font>
- <font size = "3">  **Next: [Atomic Form Factor](CH2_02-Atomic_Form_Factor.ipynb)** </font>
- <font size = "3">  **Chapter 2: [Diffraction](CH2_00-Diffraction.ipynb)** </font>
- <font size = "3">  **List of Content: [Front](../_MSE672_Intro_TEM.ipynb)** </font>